server/api/src/rsDataObjGet.cpp

```diff
@@ -284,13 +284,32 @@ int rsDataObjGet(
     std::memset(*portalOprOut, 0, sizeof(portalOprOut_t));
 
     try {
-        const auto replica_size = L1desc[fd].dataObjInfo->dataSize;
-        if (const auto single_buffer_size = irods::get_advanced_setting<const int>(irods::KW_CFG_MAX_SIZE_FOR_SINGLE_BUFFER) * 1024 * 1024;
-            replica_size <= single_buffer_size && UNKNOWN_FILE_SZ != replica_size)
+        constexpr int minimum_value_for_max_single_buffer_size_in_mb = 1;
+        constexpr int maximum_value_for_max_single_buffer_size_in_mb = 2047;
+        const auto buffer_size_in_mb = irods::get_advanced_setting<const int>(irods::KW_CFG_MAX_SIZE_FOR_SINGLE_BUFFER);
+
+        // The buffer size is converted to bytes and the value is used in malloc and memset calls below. We need to make
+        // sure that the value is valid. The schema enforces a minimum value of 1, but very large values can overflow to
+        // an invalid value. The minimum value to cause an overflow is 2048, so it must be in the range [1, 2047].
+        if (buffer_size_in_mb < minimum_value_for_max_single_buffer_size_in_mb ||
+            buffer_size_in_mb > maximum_value_for_max_single_buffer_size_in_mb)
         {
-            dataObjOutBBuf->buf = std::malloc(single_buffer_size);
-            std::memset(dataObjOutBBuf, 0, sizeof(single_buffer_size));
-            dataObjOutBBuf->len = single_buffer_size;
+            THROW(CONFIGURATION_ERROR,
+                  fmt::format("{}: [{}] configuration value is invalid. Must be between {} and {}, inclusive.",
+                              __func__,
+                              irods::KW_CFG_MAX_SIZE_FOR_SINGLE_BUFFER,
+                              minimum_value_for_max_single_buffer_size_in_mb,
+                              maximum_value_for_max_single_buffer_size_in_mb));
+        }
```

Comment:
Seems this could be replaced by adding more constraints on the property within the json schema?

https://github.com/irods/irods/blob/86eae2f46342a780563b34f4874e976f4847a879/schemas/configuration/v5/server_config.json.in#L75-L78

Reply:
I think we can safely lean on the schema. There are other situations like this in the server. To test those situations, we lean on the config validation check (`irodsServer -c`).

https://github.com/irods/irods/blob/86eae2f46342a780563b34f4874e976f4847a879/scripts/irods/test/test_misc.py#L479-L510

As long as we can guarantee that config validation is performed before the server takes any changes, we're good.

Reply:
Okay. I will replace the existing tests and remove the new checks.

Reply:
The check for the buffer size range has been removed in favor of leaning on the configuration schema. Additionally, configuration validation tests have replaced the former tests.